pandas/core/frame.py

```diff
@@ -4120,6 +4120,38 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
         if not isinstance(keys, list):
             keys = [keys]
 
+        err_msg = ('The parameter "keys" may be a column key, one-dimensional '
+                   'array, or a list containing only valid column keys and '
+                   'one-dimensional arrays.')
+
+        missing = []
+        for col in keys:
+            if isinstance(col, (ABCIndexClass, ABCSeries, np.ndarray, list)):
+                # arrays are fine as long as they are one-dimensional
+                if getattr(col, 'ndim', 1) > 1:
+                    raise ValueError(err_msg)
+            elif is_list_like(col) and not hasattr(col, '__len__'):
```

Comment:
An alterative (in light of the general comment I made), an alternative might be to only specifically check for `isinstance(col, collections.abc.Iterator`) ? That should catch the common cases you mentioned like map, iter, generator, without the risk of catching things we didn't mean to catch.

BTW, one example that still passes here is a range object (so `df.set_index(range(n))` raises a "KeyError: "None of [Int64Index([0, 1, 2], dtype='int64')] are in the [columns]"", because it is seen as a missing key. 
That might be another case for which to add a specific check here (`or isinstance(col, range)`). We could also consider actually allowing range objects (converting it to an array), but that discussion is maybe better for another issue :-)